src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,456 @@
+                {
+                    // A required delimiter is missing; do not treat semicolon as statement completion
+                    nextCommandHandler();
+                    return;
+                }
+
+                if (currentNode.Parent == null)
+                {
+                    nextCommandHandler();
+                    return;
+                }
+
+                currentNode = currentNode.Parent;
+            }
+
+            // if the statement syntax itself requires a closing delimeter, verify it is there
+            if (finalDelimiterNeedsSemicolon)
+            {
+                if (!StatementClosingDelimiterExists(currentNode, ref lastDelimiterSpan))
+                {
+                    // Example: missing final `)` in `do { } while (x$$`
+                    nextCommandHandler();
+                    return;
+                }
+            }
+
+            // if you haven't found any enclosures, put semicolon at end of statement
+            if (lastDelimiterSpan == default)
+            {
+                lastDelimiterSpan = currentNode.Span;
+            }
+
+            // Move to space after the last delimiter
+            args.TextView.TryMoveCaretToAndEnsureVisible(args.SubjectBuffer.CurrentSnapshot.GetPoint(GetEndPosition(root, lastDelimiterSpan.End, currentNode.Kind())));
+            nextCommandHandler();
+        }
+
+        private static SyntaxNode GetEnclosingArgumentList(SyntaxNode currentNode, ISyntaxFactsService syntaxFacts)
+        {
+            while (currentNode.Kind() != SyntaxKind.ArgumentList
+                && currentNode.Kind() != SyntaxKind.ArrayRankSpecifier)
+            {
+                if (currentNode.Kind() == SyntaxKind.InterpolatedStringExpression || currentNode.Kind() == SyntaxKind.StringLiteralExpression)
+                {
+                    return null;
+                }
+
+                if (currentNode == null
+                    || syntaxFacts.IsStatement(currentNode)
+                    || currentNode.Kind() == SyntaxKind.VariableDeclaration)
+                {
+                    return null;
+                }
+
+                currentNode = currentNode.Parent;
+                if (currentNode == null)
+                {
+                    return null;
+                }
+            }
+
+            // now we're in an argument list, so return the enclosing statement
+            while (!syntaxFacts.IsStatement(currentNode) && !(currentNode.Kind() == SyntaxKind.VariableDeclaration))
+            {
+                currentNode = currentNode.Parent;
+                if (currentNode == null)
+                {
+                    return null;
+                }
+            }
+
+            return currentNode;
+        }
+
+        private static bool IsCaretAtEndOfLine(SnapshotPoint caret)
+        {
+            return caret.Position == caret.GetContainingLine().End;
```

Comment:
i think anything we do should work even if there is whitespace after the caret.

Reply:
➡️ This is a helper for `GetToken` below, so I'll continue the discussion there.